If 110 dev ops write ansible role for certbot

henry22 opened the pull request:
## Description
This PR implements a new Ansible role for Certbot that automates SSL certificate management and renewal. The role provides a robust solution for managing Let's Encrypt certificates with automated renewal capabilities through systemd timers.

## Features Implemented
✅ Directory structure for letsencrypt and renewal hooks
✅ Systemd service and timer files
✅ Proper permissions and ownership
✅ All tests passing
✅ Clean template-based configuration

## Key Components
1. **Directory Structure**
   - Created `/etc/letsencrypt` base directory
   - Implemented certbot.service for renewal execution
   - Set appropriate permissions and ownership
2. **Systemd Configuration**
   - Added certbot.timer for automated daily renewal
   - Implemented certbot.service for renewal execution
   - Configured with proper security settings
3. **Testing**
   - Comprehensive test suite implemented
   - All assertions passing
   - Container-friendly test environment

## Testing Done
- Tested in Docker container environment
- Verified directory creation and permissions
- Validated systemd configuration
- Confirmed automated renewal setup

## Documentation
[https://www.notion.so/Certbot-Role-Documentation-f4a377defdd54b71b1c2e9d241ee469e?pvs=4](https://www.notion.so/Certbot-Role-Documentation-f4a377defdd54b71b1c2e9d241ee469e?pvs=4)

## Checklist
- [x] Code follows Ansible best practices
- [x] Tests are passing
- [x] Documentation updated
- [x] Template files included
- [x] Proper error handling implemented

henry22 committed:
feat: Add molecule testing configuration for certbot role

- Added molecule.yml with Docker configuration
- Created Dockerfile for test environment
- Configured test instance with required packages
- Set up systemd support in container
- Added necessary Python dependencies

henry22 committed:
feat(certbot): implement certbot role with automated renewal

✅ Directory structure for letsencrypt and renewal hooks
✅ Systemd service and timer files
✅ Proper permissions and ownership
✅ All tests passing
✅ Clean template-based configuration

- Created directory structure for letsencrypt certificates
- Implemented renewal hooks (pre, post, deploy)
- Added systemd timer for automated certificate renewal
- Added systemd service for certbot renewal
- Configured proper file permissions and ownership
- Added comprehensive tests with all assertions passing
- Used template-based configuration for maintainability

markma85 commented on roles/certbot/handlers/main.yml:
Why restart jenkins service here? I think certibot role should not depend on jenkins role.

henry22 replied:
Right! The certbot role should be independent and not have direct dependencies on Jenkins.

markma85 commented on roles/certbot/tasks/main.yml:
Systemd is not available in container, please see https://stackoverflow.com/questions/51979553/is-it-recommended-to-run-systemd-inside-docker-container

Instead, can use a separate /etc/init.d/jenkins for legacy service commands.

henry22 replied:
Modify the tasks to handle both systemd and init.d environments.

markma85 commented on roles/certbot/tests/test-certbot.yml:
```diff
@@ -0,0 +1,258 @@
+    var: apt_update_result
+
+- name: Debug certbot installation attempt
+  ansible.builtin.command: apt-get install -y certbot
```
ansible.builtin.apt has the same function.
Besides, in tests/main.yml, already included the role tasks, the package should be installed in there, here in test you only need to verify if the package is properly installed, present on the apt package list.

henry22 replied:
Right! Remove the direct `apt-get` install command since we're already handling package installation properly through `ansible.builtin.apt`.

markma85 commented on roles/certbot/tests/inventory.yml:
```diff
@@ -0,0 +1,12 @@
+          ansible_host: localhost
+          ansible_port: 23
+          ansible_user: root
+          remote_tmp: /var/tmp
```
where referenced these 2 attributes? remote_tmp, ansible_remote_tmp

henry22 replied:
Thanks, remove unused attributes.

markma85 commented on roles/certbot/tests/inventory:
can be removed since you are using the inventory.yml

henry22 replied:
Remove plain 'inventory' file.

markma85 commented on roles/certbot/defaults/main.yml:
```diff
@@ -0,0 +1,14 @@
+certbot_state: "California"
+certbot_locality: "San Francisco"
+certbot_organization: "Jenkins CI/CD"
+certbot_common_name: "{{ ansible_fqdn | default(ansible_host) | default('localhost') }}"
```
where referenced these variables?

henry22 replied:
Remove unused variables and update paths to use variables.

henry22 committed:
refactor(certbot): update role variables and defaults

- Remove unused certificate variables
- Add certbot_cert_dir variable
- Add certbot_renewal_hooks_dir variable
- Keep certbot_common_name variable
- Update variable documentation

henry22 committed:
feat(certbot): implement variable-based paths and container detection

- Use certbot_cert_dir for certificate directory
- Use certbot_renewal_hooks_dir for hooks
- Add container detection for systemd operations
- Skip systemd operations in containers
- Update directory creation tasks with variables

henry22 committed:
feat(certbot): update service template with common name

- Add certbot_common_name to service template
- Update ExecStart command with variable
- Keep timer template unchanged

henry22 committed:
test(certbot): update test suite for container environment

- Add variable definitions to playbook
- Update assertions for file existence
- Remove systemd status checks in containers
- Add debug output for troubleshooting
- Fix assertion syntax for systemd files

henry22 committed:
refactor(certbot): remove jenkins dependency

- Remove jenkins restart handler
- Make certbot role independent
- Follow single responsibility principle
- Improve role reusability

henry22 committed:
feat(certbot): add init.d script for container environments

- Create certbot-renew.sh.j2 template
- Add proper init.d header information
- Implement renewal command handling
- Use certbot_common_name variable
- Add executable permissions

henry22 committed:
feat(certbot): add container-specific configuration

- Add container environment detection
- Install cron package in containers
- Create init.d script for non-systemd environments
- Set up daily cron job for certificate renewal
- Keep systemd configuration for non-container environments

henry22 committed:
test(certbot): update tests for container compatibility

- Add environment-aware assertions
- Verify both systemd and init.d configurations
- Add container detection in tests
- Update test messages
- Ensure all configurations are properly verified

henry22 committed:
test(certbot): improve package verification tests

- Remove redundant apt-get install commands
- Use package_facts for package verification
- Add comprehensive package availability checks
- Improve error messages with detailed package info
- Clean up test assertions

henry22 committed:
refactor(certbot): clean up test inventory and remove unused tmp directory attributes

henry22 committed:
refactor(certbot): remove redundant inventory file

markma85 commented on roles/certbot/tests/ansible-playbook.yml:
```diff
@@ -0,0 +1,57 @@
+    certbot_renewal_hooks_dir: /etc/letsencrypt/renewal-hooks
+    certbot_common_name: "{{ ansible_fqdn | default(ansible_host) | default('localhost') }}"
+  pre_tasks:
+    - name: Create tmp directory
+      ansible.builtin.file:
+        path: /var/tmp
+        state: directory
+        mode: "1777"
+      become: yes
```
Could you explain this task?

henry22 replied:
The SSL certificate manager requires a temporary directory to store files during the certificate request and renewal process.

markma85 reviewed (approved):
Looks good to me.
